docs/developing.md

```diff
@@ -0,0 +1,309 @@
+    The `ParameterAttribute` and `IndexedParameterAttribute` classes offer considerable flexibility for validating inputs.
+    Defining these attributes at the class level implements the corresponding behavior in the default `__init__` function.
+    - Class-level definitions `_MAX_SUPPORTED_SECTION_VERSION` and `_MAX_SUPPORTED_SECTION_VERSION`.
+    ParameterHandler versions allow us to evolve ParameterHandler behavior in a controlled, recorded way.
+    Force field development is experimental by nature, and it is unlikely that the initial choice of header attributes is suitable for all use cases.
+    Recording the "versions" of a SMIRNOFF spec tag allows us to encode the default behavior and API of a specific generation of ParameterHandlers, while allowing the safe addition of new attributes and behaviors.
+
+  Each ParameterHandler-derived class MAY implement:
+    - `known_kwargs`: Keyword arguments passed to `ForceField.create_openmm_system` are validated against the `known_kwargs` lists of each ParameterHandler that the ForceField owns.
+      If present, these kwargs and their values will be passed on to the ParameterHandler.
+    - `to_dict`: converts the ParameterHandler to a hierarchical dict compliant with the SMIRNOFF specification.
+      The default implementation of this function should suffice for most developers.
+    - `check_handler_compatibility`: Checks whether this ParameterHandler is "compatible" with another.
+      This function is used when a ForceField is attempted to be constructed from *multiple* SMIRNOFF data sources, and it is necessary to check that two sections with the same tagname can be combined in a sane way.
+      For example, if the user instructed two `vdW` sections to be read, but the sections defined different vdW potentials, then this function should raise an Exception indicating that there is no safe way to combine the parameters.
+      The default implementation of this function should suffice for most developers.
+    - `postprocess_system`: operates identically to `create_force`, but is run after each ParameterHandlers' `create_force` has already been called.
+      The default implementation of this method simply does nothing, and should suffice for most developers.
+
+[`ParameterAttribute`](openff.toolkit.typing.engines.smirnoff.parameters.ParameterAttribute)
+: A single parameter value with units and runtime validation.
+
+[`IndexedParameterAttribute`](openff.toolkit.typing.engines.smirnoff.parameters.IndexedParameterAttribute)
+: A variable number of parameter values with units and runtime validation.
+For example, dihedral torsions are often parameterized as the sum of several sine wave terms.
+Each of the parameters of the sine wave `k`, `periodicity`, and `phase` are implemented as `IndexedParameterAttribute`s.
```

Comment:
```suggestion
[`IndexedParameterAttribute`](openff.toolkit.typing.engines.smirnoff.parameters.IndexedParameterAttribute)
: A ParameterAttribute that may appear with multiple integer values. The exact name of an IndexedParameterAttribute is NOT expected to appear verbatim in a OFFXML file, but instead should appear with a numerical integer suffix (for example the IndexedParameterAttribute `k` should only appear as, `k1`, `k2`... in an OFFXML). The current implementation requires this indexing to start at 1 and subsequent values be contiguous (no skipping numbers), but does not enforce an upper limit on the integer. 

Old version:
A variable number of parameter values with units and runtime validation.
For example, dihedral torsions are often parameterized as the sum of several sine wave terms.
Each of the parameters of the sine wave `k`, `periodicity`, and `phase` are implemented as `IndexedParameterAttribute`s.
```

Reply:
Incorporated these suggestions in dbcc59a